src/fp/ops.rs

```diff
@@ -0,0 +1,417 @@
+/// GF(p) for which the prime modulus `p` fits in one word of
+/// `W ≤ 128` bits.
+pub trait FieldParameters<W: Word> {
+    /// The prime modulus `p`.
+    const PRIME: W;
+    /// `mu = -p^(-1) mod 2^LOG2_BASE`.
+    const MU: W;
+    /// `r2 = R^2 mod p`.
+    const R2: W;
+    /// The `2^num_roots`-th -principal root of unity. This element
+    /// is used to generate the elements of `roots`.
+    const G: W;
+    /// The number of principal roots of unity in `roots`.
+    const NUM_ROOTS: usize;
+    /// Equal to `2^b - 1`, where `b` is the length of `p` in bits.
+    const BIT_MASK: W;
+    /// `roots[l]` is the `2^l`-th principal root of unity, i.e.,
+    /// `roots[l]` has order `2^l` in the multiplicative group.
+    /// `roots[0]` is equal to one by definition.
+    const ROOTS: [W; MAX_ROOTS + 1];
+    /// The log2(base) for the base used for multiprecision arithmetic.
+    /// So, `LOG2_BASE ≤ 64` as processors have at most a 64-bit
+    /// integer multiplier.
+    #[cfg(test)]
+    const LOG2_BASE: usize;
+    /// The log2(R) where R is the machine word-friendly modulus
+    /// used in the Montgomery representation.
+    #[cfg(test)]
+    const LOG2_RADIX: usize;
+}
+
+/// `FieldOps` provides arithmetic operations over GF(p).
+///
+/// The multiplication method is required as it admits different
+/// implementations.
+pub trait FieldOps<W: Word>: FieldParameters<W> {
+    /// Addition. The result will be in [0, p), so long as both x
+    /// and y are as well.
+    #[inline(always)]
+    fn add(x: W, y: W) -> W {
+        //   0,x
+        // + 0,y
+        // =====
+        //   c,z
+        let (z, carry) = x.overflowing_add(&y);
+
+        //     c, z
+        // -   0, p
+        // ========
+        // b1,s1,s0
+        let (s0, b0) = z.overflowing_sub(&Self::PRIME);
+        let (_s1, b1) = W::Carry::from(carry)
+            .as_()
+            .overflowing_sub(&W::Carry::from(b0).as_());
+        // if b1 == 1: return z
+        // else:       return s0
+        let mask = W::ZERO.wrapping_sub(&W::Carry::from(b1).as_());
+        (z & mask) | (s0 & !mask)
+    }
+
+    /// Subtraction. The result will be in [0, p), so long as both x
+    /// and y are as well.
+    #[inline(always)]
+    fn sub(x: W, y: W) -> W {
+        //        x
+        // -      y
+        // ========
+        //    b0,z0
+        let (z0, b0) = x.overflowing_sub(&y);
+        let mask = W::ZERO.wrapping_sub(&W::Carry::from(b0).as_());
+        //      z0
+        // +     p
+        // ========
+        //   s1,s0
+        z0.wrapping_add(&(mask & Self::PRIME))
+        // z0 + (m & self.p)
+        // if b1 == 1: return s0
+        // else:       return z0
+    }
+
+    /// Negation, i.e., `-x (mod p)` where `p` is the modulus.
+    #[inline(always)]
+    fn neg(x: W) -> W {
+        Self::sub(W::ZERO, x)
+    }
+
+    #[inline(always)]
+    fn modp(x: W) -> W {
+        Self::sub(x, Self::PRIME)
```

Comment:
Correct, this works since prime modulus has the form `p=2^W-x`, where `W` is the word size. 
I was thinking how to make this operation more general, if this is ever needed.